### PR TITLE
SPH-598 SPH-840 - Feature/correct interpretation mozaiekvegetaties

### PR DESCRIPTION
**Pull request by JRdeLange**
Wederom van boven naar beneden per bestand kort wat er gedaan is :)

- `mozaiekjson.json` / `opgeschoonde_definitietabel.xlsx`
  - habtype -> kwalificerend_habtype en alleen_zelfstandig -> ook_mozaiekvegetaties
- `functionality_test.ipynb`
  - Enkel 2 cellen omgewisseld
- `test_deftabel_matching.py` / `test_env_vars_overwrite_config.py` / `test_mozaiek_geometric_operations.py`
  - Aangepast aan de nieuwe mozaiekregeldingen
- `test_mozaiek_initialization_and_evaluation`
  - Grondig omgegooid zodat hij de nieuwe functionaliteit ook goed test. Kijk vooral even of je nog test/edge cases mist
- `definitietabel.py`
  - Kwalificierende vegtypen bepalen aanroepen in mozaiekregels, en we strippen whitespace uit Habitattype in de deftabel want er was er 1 die 'H1234 ' had ipv "H1234" wat problematisch was voor een test (en gewoon niet netjes)
- `enums.py`
  - Kleine netheiddingetjes
- `habitat.py`
  - Inconsistentie gefixt in de output van de mits/mozaiekopmerkingen
- `common.py`
  - Nieuwe constante voor de mozaiekregels
- `mozaiek.py` / `vegkartering.py`
  - Mozaiekregels checken werkt nu (in het kort) zo:
    1. In `vegkartering.bepaal_mozaiek_habitatkeuzes()` maken we met `mozaiek.make_buffered_boundary_overlay_gdf()` een df met daarin voor ieder mozaiekvlak alle benodigde data per omringend vlak
    2. Met `mozaiek.construct_elmid_omringd_door_gdf()` wordt dit verder uitgewerkt, zodat er voor ieder mozaiekvlak een regel is voor ieder complexdeel wat er omheen ligt
    3. Het voor ieder vlak relevante deel hiervan wordt dan in `StandaardMozaiekregel.check()` gevoerd om te kijken of er voldaan wordt aan de mozaiekregel
       1. Hier wordt eerst `self._vul_mozk_perc_dict()` gedaan om uiteindelijk het `_Mozk_info{i}` veld mee te vullen. Deze is gescheiden van volgende stappen voornamelijk om code wat strakker en meer single purpose te houden, maar zou evt voor optimalisaties nog gecombineerd kunnen worden met de volgende stap
       2. Daarna wordt de omringing door zowel kwalificerende vlakken als HXXXX vlakken van het mozaiekvlak bepaald met `self._bepaal_kwalificerende_en_HXXXX_omringing()`
       3. Op basis van deze 2 percentages wordt dan een truth value bepaald

### CODE REVIEW

File: veg2hab/definitietabel.py [hunk 1]
```diff
@@ -46,11 +46,11 @@ def __init__(self, df: pd.DataFrame):
             .apply(MozaiekRegel.parse_raw)
         )
         # Aanmaken dict keys die gebruikt gaan worden om de mozaiekregels te checken
-        # TODO: Om deze isinstance heenwerken voor modulariteit
-        # TODO: Idealiter komt dit een een soort post_init (https://stackoverflow.com/questions/66571079/alter-field-after-instantiation-in-pydantic-basemodel-class)
```
Comment (mark-boer) [inline veg2hab/definitietabel.py:50]: TODO mag blijven staan toch?

File: veg2hab/mozaiek.py [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
 
 import geopandas as gpd
+import pandas as pd
 from pydantic import BaseModel, Field, PrivateAttr
 
 from veg2hab.enums import Kwaliteit, MaybeBoolean
 from veg2hab.io.common import Interface
+from veg2hab.vegetatietypen import SBB, VvN
+
+MozkPercTuple = namedtuple("MozkPercTuple", ["habtype", "kwaliteit", "percentage"])
```
Comment (mark-boer) [inline veg2hab/mozaiek.py:14]: Is NamedTuple uit Typing al onderdeel van python3.7?

File: veg2hab/mozaiek.py [hunk 3]
```diff
@@ -93,76 +104,301 @@ def __str__(self):
...
+            ElmID | habtype | kwaliteit | vegtypen | complexdeel_percentage | omringing_percentage
+            Er is een rij voor ieder complexdeel in de omliggende vlakken.
 
-    def check(self, habtype_percentage_dict: Dict) -> None:
-        requested_habtype_percentage = 0
-        for key in self.keys:
-            requested_habtype_percentage += habtype_percentage_dict.get(key, 0)
+        Vult ook de tegengekomen_kwal_vegtypen en mozk_perc_dict in.
+        """
+        assert all(
```
Comment (mark-boer) [inline veg2hab/mozaiek.py:158]: je kunt ook set comparisons doen hier. `issubset` of `issuperset` of `<=`

File: veg2hab/mozaiek.py [hunk 4]
```diff
@@ -93,76 +104,301 @@ def __str__(self):
...
 
-        # Threshold kan niet meer behaald worden, dus FALSE
         self.cached_evaluation = MaybeBoolean.FALSE
 
-    def __str__(self):
-        complete_string = ""
-        complete_string += f"{'(als rand langs)' if self.ook_als_rand_langs else ''} "
-        complete_string += (
-            f"{'zelfstndg' if self.alleen_zelfstandig else 'zelfstndg/mozk'} "
+    def _vul_mozk_perc_dict(self, omringd_door: pd.DataFrame) -> None:
```
Comment (mark-boer) [inline veg2hab/mozaiek.py:202]: Ik heb een lichte voorkeur voor functies die gewoon de waarde teruggeven, i.p.v. de staat van de class aan te passen, maakt het iets makkelijker te testen. Maar is ook prima hoor.

File: veg2hab/mozaiek.py [hunk 5]
```diff
@@ -93,76 +104,301 @@ def __str__(self):
...
+            ), f"Omringing percentage is niet hetzelfde voor alle complexdelen in vlak met ElmID {vlak_group.ElmID.iloc[0]}"
+
+            grouped_by_habtype = vlak_group.groupby("habtype")
+            for habtype, habtype_group in grouped_by_habtype:
+                bedekking_goed_habtype = habtype_group[
+                    habtype_group.kwaliteit == Kwaliteit.GOED
+                ].complexdeel_percentage.sum()
+
+                if bedekking_goed_habtype >= self.mozaiek_minimum_bedekking:
+                    mozk_perc_dict[(habtype, Kwaliteit.GOED)] += habtype_group.iloc[
```
Comment (mark-boer) [inline veg2hab/mozaiek.py:224]: wat doet deze iloc[0]?

File: veg2hab/mozaiek.py [hunk 6]
```diff
@@ -93,76 +104,301 @@ def __str__(self):
...
+                > self.mozaiek_minimum_bedekking
+            ):
+                omringing_HXXXX += vlak_group.iloc[0].omringing_percentage
+                continue
+
+        return (omringing_kwal_vlakken, omringing_HXXXX, tegengekomen_kwal_vegtypen)
+
+    def _bepaal_kwalificerende_en_HXXXX_bedekking(
+        self, vlak_group: pd.DataFrame
+    ) -> Tuple[Union[int, float], Union[int, float], Set[Union[SBB, VvN]]]:
```
Comment (mark-boer) [inline veg2hab/mozaiek.py:300]: Als je vaak dezelfde Union gebruikt, kun je ervoor kiezen om deze aan een variabele toe te kennen:

```
Number = Union[Int, Float]
```

File: veg2hab/mozaiek.py [hunk 7]
```diff
@@ -93,76 +104,301 @@ def __str__(self):
...
+
+        for row in vlak_group.itertuples():
+            if (
+                # Als het habitattype matcht en er wordt aan de kwaliteitseisen voldaan
+                row.habtype == self.kwalificerend_habtype
+                and (row.kwaliteit == Kwaliteit.GOED or not self.alleen_goede_kwaliteit)
+            ) or (
+                # Of als de mozaiekvegetaties ook toegestaan zijn en daar is een match
+                self.ook_mozaiekvegetaties
+                and any(
```
Comment (mark-boer) [inline veg2hab/mozaiek.py:314]: In theorie zijn dit allemaal set operaties.. al is dit wel lekker leesbaar zo :-)

File: veg2hab/mozaiek.py [hunk 8]
```diff
@@ -93,76 +104,301 @@ def __str__(self):
...
+                    vegtype in self.kwalificerende_vegtypen for vegtype in row.vegtypen
+                )
+            ):
+                # Dan tellen we het percentage van het complexdeel mee
+                bedekking_kwal_complexdelen += row.complexdeel_percentage
+                tegengekomen_kwal_vegtypen.update(
+                    [
```
Comment (mark-boer) [inline veg2hab/mozaiek.py:321]: dit is een set.intersection, maar weet niet of het er leesbaarder van wordt :-)

Geeft mogelijk wel een performance winst, als dit langzaam blijkt te zijn.

File: veg2hab/mozaiek.py [hunk 9]
```diff
@@ -93,76 +104,301 @@ def __str__(self):
...
+    ):
+        # Casten naar list voor (de)serializatie
+        self.tegengekomen_kwal_SBB = list(
+            {vegtype for vegtype in tegengekomen_vegtypen if isinstance(vegtype, SBB)}
         )
+        self.tegengekomen_kwal_VvN = list(
+            {vegtype for vegtype in tegengekomen_vegtypen if isinstance(vegtype, VvN)}
+        )
+
+    def _bepaal_kwalificerende_en_HXXXX_omringing(
```
Comment (mark-boer) [inline veg2hab/mozaiek.py:260]: docstring hier kan geen kwaad, code is moeilijk te begrijpen.

File: veg2hab/mozaiek.py (hunk 6 above)
Comment (mark-boer) [inline veg2hab/mozaiek.py:300]: +docstrings.

**Review by mark-boer — APPROVED** [review]
Zie comments:

Een aantal functies kunnen zeker nog een docstrings gebruiken. Verder ziet het er wel goed uit.

Het is niet de makkelijkste code om te volgen wat er gebeurd.

File: veg2hab/io/common.py [hunk 10]
```diff
@@ -155,9 +155,13 @@ class Config:
         description="Threshold voor het bepalen of een vlak in het mozaiek ligt",
     )
     mozaiek_als_rand_threshold: Union[int, float] = Field(
-        default=50.0,
+        default=95.0,
```
[the remark this answers is not in the document]
Comment (JRdeLange) [reply veg2hab/io/common.py]: Had ik ook net gespot https://github.com/Spheer-ai/veg2hab/pull/63/commits/6c147d3bb9e4a235deaa13b7112c6ead66a5678c

File: veg2hab/mozaiek.py (hunk 2 above)
Comment (JRdeLange) [reply veg2hab/mozaiek.py:14]: Ik draai 3.7 dus denk het wel dan

File: veg2hab/definitietabel.py (hunk 1 above)
Comment (JRdeLange) [reply veg2hab/definitietabel.py:50]: Ja kan ook wel idd

File: veg2hab/definitietabel.py [hunk 11]
```diff
@@ -46,11 +46,11 @@ def __init__(self, df: pd.DataFrame):
...
         self.df["Mozaiekregel"].apply(
-            lambda regel: regel.determine_keys()
-            if isinstance(regel, StandaardMozaiekregel)
+            lambda regel: regel.determine_kwalificerende_vegtypen(
+                self.df[self.df.Habitattype == regel.kwalificerend_habtype]
+            )
+            if isinstance(regel, StandaardMozaiekregel) and regel.ook_mozaiekvegetaties
```
[the remark this answers is not in the document]
Comment (JRdeLange) [reply veg2hab/definitietabel.py:53]: Ik had het zo gedaan om het onnodig slicen en doorgeven van de dataframe te voorkomen, maar zeg maar wat je liever hebt :)

File: veg2hab/mozaiek.py (hunk 4 above)
Comment (JRdeLange) [reply veg2hab/mozaiek.py:202]: Ben ik met je eens, heb het hier zo gelaten omdat de assigments wat ingewikkelder zijn dan self.field = waarde, en ik wilde die logica uit check() houden.

File: veg2hab/mozaiek.py (hunk 6 above)
Comment (JRdeLange) [reply veg2hab/mozaiek.py:300]: Oh nice :)

File: veg2hab/mozaiek.py (hunk 7 above)
Comment (JRdeLange) [reply veg2hab/mozaiek.py:314]: Is het veel sneller met set operaties? Comprehension is ook best vlot toch.

File: veg2hab/mozaiek.py (hunk 8 above)
Comment (JRdeLange) [reply veg2hab/mozaiek.py:321]: Zet er wel even een NOTE comment bij dat dit een kandidaat is voor optimalisatie.

File: veg2hab/mozaiek.py (hunk 9 above)
Comment (JRdeLange) [reply veg2hab/mozaiek.py:260]: Het is idd nogal gegoochel met percentages, thresholds en truth values. Hopelijk is het nu duidelijker, zo niet hoor ik t wel.

**Comment by mark-boer** [comment]
Ja veel beter, ik vind het zelf vaak nog duidelijker worden als er even een klein voorbeeldje staat van data in en data uit. Maar dit is prima zo!
